### PR TITLE
feat(jupyter): Add display support for layouts

### PR DESCRIPTION
**Pull request by jourdain**

### CODE REVIEW

File: trame_client/ui/core.py [hunk 1]
```diff
@@ -8,12 +15,23 @@ class AbstractLayout:
...
 
-    def __init__(self, _server, _root_elem, template_name="main", **kwargs):
+    def __init__(
+        self,
+        _server,
+        _root_elem,
+        template_name="main",
+        width="100%",
+        height="600px",
+        base_url="http://localhost",
```
Comment (banesullivan) [inline trame_client/ui/core.py:25]: I can't remember exactly why at the moment, but I remember `127.0.0.1` being more robust than `localhost`

```suggestion
        base_url="http://127.0.0.1",
```

File: trame_client/ui/core.py [hunk 2]
```diff
@@ -66,3 +98,36 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_value, exc_traceback):
         self._original_root.__exit__(exc_type, exc_value, exc_traceback)
         self.flush_content()
+
+    # -------------------------------------------------------------------------
+    # Jupyter display integration
+    # -------------------------------------------------------------------------
+
+    def __jupyter_content(self):
```
Comment (banesullivan) [inline trame_client/ui/core.py:106]: Let's not name-mangle this and just have one underscore -- for the off chance we want to use this somewhere else:

```suggestion
    def _jupyter_content(self):
```

(needs to be changed where used too)

**Review by banesullivan — APPROVED** [review]
Excellent start! This is pretty nifty in playing with this locally.

The `display_html` fallback is a good-catch if `ipywidgets` is unavailable.

Inline suggestions but nothing critical

File: trame_client/ui/core.py (hunk 1 above)
Comment (banesullivan) [reply trame_client/ui/core.py:25]: Or this `base_url` is client-side? Hm. `localhost` *should* be fine

Comment (banesullivan) [reply trame_client/ui/core.py:25]: Ignore this

**Comment by jourdain** [comment]
Let me know if you are good for me to merge it...

File: trame_client/ui/core.py [hunk 3]
```diff
@@ -66,3 +98,35 @@ def __enter__(self):
...
+                    </div>"""
+
+        elem_id = f"{self.server.name}_{self._template_name}"
+        src = f"{self.iframe_base_url}:{self.server.port}/index.html?ui={self._template_name[16:]}"
+        return (
+            f'<iframe id="{elem_id}" src="{src}" style="{self.iframe_style}"></iframe>'
+        )
+
+    @property
+    def widget(self):
```
Comment (jourdain) [inline trame_client/ui/core.py:121]: @banesullivan I'm just worried about that name. Should this be `ipywidget` instead to make it explicit?

**Comment by jourdain** [comment]
Usage example

```python
from trame.app import get_server, jupyter
from trame.ui.vuetify import SinglePageLayout
from trame.widgets import vuetify, vtk as vtk_widgets

server = get_server()
state, ctrl = server.state, server.controller

with SinglePageLayout(server) as layout:
    with layout.content:
        with vuetify.VContainer(fluid=True, classes="pa-0 fill-height"):
            with vtk_widgets.VtkView() as view:
                ctrl.view_reset_camera = view.reset_camera
                with vtk_widgets.VtkGeometryRepresentation():
                    vtk_widgets.VtkAlgorithm(
                        vtk_class="vtkConeSource",
                        state=("{ resolution }",),
                    )

    with layout.toolbar:
        vuetify.VSpacer()
        vuetify.VSlider(
            v_model=("resolution", 6),
            min=3,
            max=60,
            step=1,
            hide_details=True,
            style="max-width: 300px;",
        )
        with vuetify.VBtn(icon=True, click=ctrl.view_reset_camera):
            vuetify.VIcon("mdi-crop-free")


await layout.ready
layout
```

```python
from trame.ui.vuetify import VAppLayout
with VAppLayout(server, template_name="control", height=60) as ui:
    vuetify.VSlider(
        v_model=("resolution", 6),
        min=3,
        max=60,
        step=1,
        hide_details=True,
        classes="my-3",
    )
ui
```

```python
import ipywidgets as widgets

def update_resolution(change):
    r = change["new"]
    with state:
        state.resolution = r

slider = widgets.IntSlider(min=3, max=60, continuous_update=True)
slider.observe(update_resolution, "value")

@state.change("resolution")
def on_change(resolution, **kwargs):
    slider.value = resolution

widgets.VBox([slider, ui.widget])
```

```python
from trame.widgets import html
not_ready = get_server("seb")
with VAppLayout(not_ready, height=60) as ui_2:
    html.Div("Hello World")

ui_2
```

```python
await ui_2.ready
ui_2
```

**Review by banesullivan — APPROVED** [review]
Pending the widget -> ipywidget change, this looks good to me!

**Comment by jourdain** [comment]
Should be all good and future proof while we will be adding more url builder.

**Review by banesullivan — APPROVED** [review]

**Pull request merged**